traffic_ops/traffic_ops_golang/topology/topologies.go

```diff
@@ -0,0 +1,485 @@
+	return util.JoinErrs(errs)
+}
+
+// Implementation of the Identifier, Validator interface functions
+func (topology TOTopology) GetKeys() (map[string]interface{}, bool) {
+	return map[string]interface{}{"name": topology.Name}, true
+}
+
+// SetKeys is a requirement of the api.Updater interface and is called by
+// api.UpdateHandler().
+func (topology *TOTopology) SetKeys(keys map[string]interface{}) {
+	topology.Name, _ = keys["name"].(string)
+}
+
+// GetAuditName is a requirement of the api.Identifier interface.
+func (topology *TOTopology) GetAuditName() string {
+	return topology.Name
+}
+
+// Create is a requirement of the api.Creator interface.
+func (topology *TOTopology) Create() (error, error, int) {
+	tx := topology.APIInfo().Tx.Tx
+	err := tx.QueryRow(insertQuery(), topology.Name, topology.Description).Scan(&topology.Name, &topology.Description, &topology.LastUpdated)
+	if err != nil {
+		return api.ParseDBError(err)
+	}
+
+	if userErr, sysErr, errCode := topology.addNodes(); userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+
+	if userErr, sysErr, errCode := topology.addParents(); userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+
+	return nil, nil, 0
+}
+
+// Read is a requirement of the api.Reader interface and is called by api.ReadHandler().
+func (topology *TOTopology) Read() ([]interface{}, error, error, int) {
+	where, orderBy, pagination, queryValues, errs := dbhelpers.BuildWhereAndOrderByAndPagination(topology.ReqInfo.Params, topology.ParamColumns())
+	if len(errs) > 0 {
+		return nil, util.JoinErrs(errs), nil, http.StatusBadRequest
+	}
+	query := selectQuery() + where + orderBy + pagination
+	rows, err := topology.ReqInfo.Tx.NamedQuery(query, queryValues)
+	if err != nil {
+		return nil, nil, errors.New("topology read: querying: " + err.Error()), http.StatusInternalServerError
+	}
+
+	var interfaces []interface{}
+	topologies := map[string]*tc.Topology{}
+	indices := map[int]int{}
+	for index := 0; rows.Next(); index++ {
+		var (
+			name, description string
+			lastUpdated       tc.TimeNoMod
+		)
+		topologyNode := tc.TopologyNode{}
+		topologyNode.Parents = []int{}
+		var parents pq.Int64Array
+		if err = rows.Scan(
+			&name,
+			&description,
+			&lastUpdated,
+			&topologyNode.Id,
+			&topologyNode.Cachegroup,
+			&parents,
+		); err != nil {
+			return nil, nil, errors.New("topology read: scanning: " + err.Error()), http.StatusInternalServerError
+		}
+		for _, id := range parents {
+			topologyNode.Parents = append(topologyNode.Parents, int(id))
+		}
+		indices[topologyNode.Id] = index
+		if _, exists := topologies[name]; !exists {
+			topology := tc.Topology{Nodes: []tc.TopologyNode{}}
+			topologies[name] = &topology
+			topology.Name = name
+			topology.Description = description
+			topology.LastUpdated = &lastUpdated
+		}
+		topologies[name].Nodes = append(topologies[name].Nodes, topologyNode)
+	}
+	if err = rows.Close(); err != nil {
```

Comment:
closing the rows needs to be in a `defer` up on L180 to ensure that the rows are properly closed however this method may exit

Reply:
This was changed in 36e085494f in an effort to fix various warnings, though I not consider cases when there is an error after the query but before finishing reading rows. Deferring again in 73fb323465.